### PR TITLE
bpo-26669: Fix nan arg value error in pytime.c

### PR DESCRIPTION
**Pull request by hahnlee**
* Make value error message for nan value
Previously, it did not show error message in arm system
* Add test case for nan value

https://bugs.python.org/issue26669

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

File: Python/pytime.c [hunk 1]
```diff
@@ -155,6 +155,10 @@ _PyTime_ObjectToTime_t(PyObject *obj, time_t *sec, _PyTime_round_t round)
         volatile double d;
 
         d = PyFloat_AsDouble(obj);
+        if (isnan(d)) {
+            PyErr_SetString(PyExc_ValueError, "Invalid Value");
```
Comment (wkschwartz) [inline Python/pytime.c]: Might be helpful for the error message specifically to mention that the value is NaN.

**Comment by pitrou** [comment]
@haypo, do you have any comments about this?

File: Lib/test/test_time.py [hunk 2]
```diff
@@ -489,6 +489,10 @@ def test_localtime_failure(self):
         self.assertRaises(OSError, time.localtime, invalid_time_t)
         self.assertRaises(OSError, time.ctime, invalid_time_t)
 
+        # Issue #26669: check for localtime() failure
+        self.assertRaises(ValueError, time.localtime, float("nan"))
+        self.assertRaises(ValueError, time.ctime, float("nan"))
```
Comment (vstinner) [inline Lib/test/test_time.py:494]: Ok for "functional" tests", but I also would like to see unit tests on the following _testcapi functions:

* pytime_object_to_time_t
* pytime_object_to_timeval
* pytime_object_to_timespec
* PyTime_FromSeconds
* PyTime_FromSecondsObject
* PyTime_AsSecondsDouble

See CPyTimeTestCase in test_time.py.

**Review by vstinner — CHANGES_REQUESTED** [review]

**Comment by pitrou** [comment]
Hmm... Did @warsaw [hack the bot](https://github.com/python/cpython/pull/3085#issuecomment-326214637)?

**Comment by vstinner** [comment]
@sn0wle0pard: Would you mind to add requested unit tests? Otherwise, I might try to modify your PR to add them myself?

**Comment by hahnlee** [comment]
@haypo I'll add test case as soon as possible.

**Comment by hahnlee** [comment]
@haypo I'm really sorry for the delay in writing the code. 
# Summary
System: Ubuntu 16.04 x86
`pytime_object_to_timeval` got error.
If the result is the same as expected, I will replace the passed test code with the changed `_PyTime_FromSecondsObject` and `_PyTime_ObjectToDenominator`. 
* pytime_object_to_time_t: pass
* pytime_object_to_timeval : *error* `assert(0 <= *usec && *usec < SEC_TO_US);`
* pytime_object_to_timespec: *error* `assert(0 <= *nsec && *nsec < SEC_TO_NS);`
* PyTime_FromSecondsObject: pass (When add some C code)
* PyTime_FromSeconds: `TypeError: integer argument expected, got float` 
* PyTime_FromSecondsObject: `TypeError: integer argument expected, got float` 

# Result
I tested the following code in `_check_rounding`.
```python
# test nan
for time_rnd, _ in ROUNDING_MODES:
    with self.assertRaises(ValueError):
        pytime_converter(float('nan'), time_rnd)
``` 
Here are the results:
```
Python/pytime.c:116: _PyTime_DoubleToDenominator: Assertion `0.0 <= floatpart && floatpart < denominator' failed.
```
...so I tested one by one.
`pytime_object_to_time_t`: pass
`pytime_object_to_timeval`: error, pytime.c:116
`pytime_object_to_timespec`: error, pytime.c:116

`PyTime_FromSeconds`: error `TypeError: integer argument expected, got float` (did not use time_rnd)
`PyTime_AsSecondsDouble`: has same error `PyTime_FromSecondsObject` (did not use time_rnd)

And `PyTime_FromSecondsObject`: did not raise
`_PyTime_FromSecondsObject` call `_PyTime_FromObject` and `_PyTime_FromObject` has same logic at `_PyTime_ObjectToTime_t` (check float but do not check nan)
So I changed `_PyTime_FromObject`  and passed test for `_PyTime_FromSecondsObject`.

In case `pytime_object_to_timeval` and `pytime_object_to_timespec` they call `_PyTime_ObjectToDenominator` and `_PyTime_ObjectToDenominator` has `PyFloat_AsDouble` too

I add `Py_IS_NAN(d)` at `_PyTime_ObjectToDenominator`

but `pytime_object_to_timeval` and `pytime_object_to_timespec`  got assert error
it mean in case `_PyTime_ObjectToDenominator`,  *obj float('nan') has problem in `usec` or `nsec`

i think `PyTime_FromSeconds` and `PyTime_AsSecondsDouble` seem reasonable.
but i'm not sure `pytime_object_to_timeval` and `pytime_object_to_timespec` has reasonable error.

If the result is the same as expected, I will replace the passed test code with the changed `_PyTime_FromSecondsObject` and `_PyTime_ObjectToDenominator`.

**Comment by hahnlee** [comment]
pytime_object_to_timeval also has same error on macOS sierra

**Comment by sn0wle0pard** [comment]
oh... in case NaN numerator is zero. i add some code for numerator zero. the it pass test case

**Review by vstinner — APPROVED** [review]

**Comment by vstinner** [comment]
Ok, it now LGTM!

**Pull request merged**